python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/sfn.py

```diff
@@ -0,0 +1,160 @@
+
+        Args:
+            context (Union[OpExecutionContext, AssetExecutionContext]): The context of the currently executing Dagster op or asset.
+            start_execution_input (Dict): Parameters for the ``start_execution`` boto3 Step Functions client call.
+            extras (Optional[Dict[str, Any]]): Additional Dagster metadata to pass to the Step Function execution.
+
+        Returns:
+            PipesClientCompletedInvocation: Wrapper containing results reported by the external
+            process.
+        """
+        params = start_execution_input
+
+        state_machine_arn = cast(str, params["stateMachineArn"])
+
+        with open_pipes_session(
+            context=context,
+            message_reader=self._message_reader,
+            context_injector=self._context_injector,
+            extras=extras,
+        ) as session:
+            _ = session.get_bootstrap_cli_arguments()
+
+            try:
+                execution_arn = self._client.start_execution(**params)["executionArn"]
+
+            except ClientError as err:
+                error_info = err.response.get("Error", {})
+                context.log.error(
+                    "Couldn't start execution %s. Here's why: %s: %s",
+                    state_machine_arn,
+                    error_info.get("Code", "Unknown error"),
+                    error_info.get("Message", "No error message available"),
+                )
+                raise
+
+            response = self._client.describe_execution(executionArn=execution_arn)
+            context.log.info(
+                f"Started AWS Step Function execution {state_machine_arn} run: {execution_arn}"
+            )
+
+            try:
+                response = self._wait_for_execution_completion(execution_arn)
+            except DagsterExecutionInterruptedError:
+                if self.forward_termination:
+                    self._terminate_execution(context=context, execution_arn=execution_arn)
+                raise
+
+            status = response["status"]
+            if status != "SUCCEEDED":
+                raise RuntimeError(
+                    f"Step Function execution {state_machine_arn} run {execution_arn} completed with status {status} :\n{response.get('errorMessage')}"
+                )
+            else:
+                context.log.info(
+                    f"Step Function execution {state_machine_arn} run {execution_arn} completed successfully"
+                )
+
+        return PipesClientCompletedInvocation(
+            session, metadata=self._extract_dagster_metadata(response)
+        )
+
+    def _wait_for_execution_completion(
+        self, execution_arn: str
+    ) -> "DescribeExecutionOutputTypeDef":
+        while True:
+            response = self._client.describe_execution(executionArn=execution_arn)
+            if response["status"] in ["FAILED", "SUCCEEDED", "TIMED_OUT", "ABORTED"]:
```

Comment:
I would prefer to do it with a waiter instead. See ECS [example](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py#L298).